### PR TITLE
Make lights_per_object configurable

### PR DESCRIPTION
**Pull request by NHodgesVFX**
you can now have lots of lights, thanks to jonbonazza for helping me with this pr

tonsoflights

some videos of it:
https://streamable.com/mjb9gw
https://streamable.com/npu07o

### CODE REVIEW

File: drivers/gles3/rasterizer_scene_gles3.cpp [hunk 1]
```diff
@@ -5168,7 +5173,7 @@ void RasterizerSceneGLES3::initialize() {
 		glBufferData(GL_UNIFORM_BUFFER, sizeof(LightDataUBO), NULL, GL_DYNAMIC_DRAW);
 		glBindBuffer(GL_UNIFORM_BUFFER, 0);
 
-		state.max_forward_lights_per_object = 8;
+		state.max_forward_lights_per_object = render_list.max_lights_per_object;
```
Comment (clayjohn) [inline drivers/gles3/rasterizer_scene_gles3.cpp:5176]: This should be ``MIN(state.max_ubo_lights, render_list.max_lights_per_object)``

The majority of devices only support a ``GL_MAX_UNIFORM_BLOCK_SIZE`` of 65536. This is true of both mobile and desktop. With a ``GL_MAX_UNIFORM_BLOCK_SIZE`` of 65536 we can have at most 65536/160 = 409 lights in the scene. A large proportion of devices only support a ``GL_MAX_UNIFORM_BLOCK_SIZE`` of 16384, which is 102 lights per scene. Since the ``max_lights_per_object`` can be so much higher, it is a good idea to limit it.

File: doc/classes/ProjectSettings.xml [hunk 2]
```diff
@@ -1073,6 +1073,9 @@
 		<member name="rendering/limits/rendering/max_renderable_reflections" type="int" setter="" getter="" default="1024">
 			Max number of reflection probes renderable in a frame. If more than this number are used, they will be ignored. On some systems (particularly web) setting this number as low as possible can increase the speed of shader compilation.
 		</member>
+		<member name="rendering/limits/rendering/max_lights_per_object" type="int" setter="" getter="" default="32">
+			Max number of lights allowed to be rendered per object. The maximum is based on the number of registers available on your hardware, if you set it to high Godot will complain and the scene will look broken.
```
Comment (clayjohn) [inline doc/classes/ProjectSettings.xml:1077]:
```suggestion
			Max number of lights renderable per object. This is further limited by hardware support. Most devices only support 409 lights, while many devices (especially mobile) only support 102. Setting this low will slightly reduce memory usage and may decrease shader compile times. 
```

File: drivers/gles3/rasterizer_scene_gles3.cpp (hunk 1 above)
Comment (NHodgesVFX) [reply drivers/gles3/rasterizer_scene_gles3.cpp:5176]: ok ill go ahead and make the changes

File: doc/classes/ProjectSettings.xml [hunk 3]
```diff
@@ -1073,6 +1073,9 @@
 		<member name="rendering/limits/rendering/max_renderable_reflections" type="int" setter="" getter="" default="1024">
 			Max number of reflection probes renderable in a frame. If more than this number are used, they will be ignored. On some systems (particularly web) setting this number as low as possible can increase the speed of shader compilation.
 		</member>
+		<member name="rendering/limits/rendering/max_lights_per_object" type="int" setter="" getter="" default="32">
+			Max number of lights renderable per object. This is further limited by hardware support. Most devices only support 409 lights, while many devices (especially mobile) only support 102. Setting this low will slightly reduce memory usage and may decrease shader compile times.
```
Comment (clayjohn) [inline doc/classes/ProjectSettings.xml:1077]: Extra space at the end of this line. oops!

File: drivers/gles3/rasterizer_scene_gles3.cpp (hunk 1 above)
Comment (NHodgesVFX) [reply drivers/gles3/rasterizer_scene_gles3.cpp]: done

**Comment by clayjohn** [comment]
Looks great!

Before your PR is ready to go, you will have to squash your 3 commits together as described here https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase

**Comment by NHodgesVFX** [comment]
> Looks great!
> 
> Before your PR is ready to go, you will have to squash your 3 commits together as described here https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase

ok done :)

**Review by clayjohn — APPROVED** [review]
Looks good to me! Great work and great first contribution!

File: drivers/gles3/rasterizer_scene_gles3.cpp [hunk 4]
```diff
@@ -1834,15 +1834,15 @@ void RasterizerSceneGLES3::_render_geometry(RenderList::Element *e) {
 
 void RasterizerSceneGLES3::_setup_light(RenderList::Element *e, const Transform &p_view_transform) {
 
-	int omni_indices[16];
+	int maxobj = state.max_forward_lights_per_object;
+
+	int *omni_indices = (int *)memalloc(maxobj * sizeof(int));
```
Comment (lawnjelly) [inline drivers/gles3/rasterizer_scene_gles3.cpp:1839]: alloca would be better here for stack allocation.

File: drivers/gles3/rasterizer_scene_gles3.cpp [hunk 5]
```diff
@@ -5054,6 +5057,8 @@ void RasterizerSceneGLES3::initialize() {
 	ProjectSettings::get_singleton()->set_custom_property_info("rendering/limits/rendering/max_renderable_lights", PropertyInfo(Variant::INT, "rendering/limits/rendering/max_renderable_lights", PROPERTY_HINT_RANGE, "16,4096,1"));
 	render_list.max_reflections = GLOBAL_DEF("rendering/limits/rendering/max_renderable_reflections", (int)RenderList::DEFAULT_MAX_REFLECTIONS);
 	ProjectSettings::get_singleton()->set_custom_property_info("rendering/limits/rendering/max_renderable_reflections", PropertyInfo(Variant::INT, "rendering/limits/rendering/max_renderable_reflections", PROPERTY_HINT_RANGE, "8,1024,1"));
+	render_list.max_lights_per_object = GLOBAL_DEF_RST("rendering/limits/rendering/max_lights_per_object", (int)RenderList::DEFAULT_MAX_LIGHTS_PER_OBJECT);
+	ProjectSettings::get_singleton()->set_custom_property_info("rendering/limits/rendering/max_lights_per_object", PropertyInfo(Variant::INT, "rendering/limits/rendering/max_lights_per_object", PROPERTY_HINT_RANGE, "1,1024,1"));
```
Comment (Calinou) [inline drivers/gles3/rasterizer_scene_gles3.cpp:5061]: Since the renderer treats values below 8 as being equal to 8, it doesn't make sense to allow values below 8:

```suggestion
	ProjectSettings::get_singleton()->set_custom_property_info("rendering/limits/rendering/max_lights_per_object", PropertyInfo(Variant::INT, "rendering/limits/rendering/max_lights_per_object", PROPERTY_HINT_RANGE, "8,1024,1"));
```

File: drivers/gles3/rasterizer_scene_gles3.cpp (hunk 4 above)
Comment (NHodgesVFX) [reply drivers/gles3/rasterizer_scene_gles3.cpp]: I made the change, thanks for the suggestion.

Comment (akien-mga) [reply drivers/gles3/rasterizer_scene_gles3.cpp]: This makes MSVC unhappy:
```
 cl /Fodrivers\gles3\rasterizer_scene_gles3.windows.opt.tools.64.obj /c drivers\gles3\rasterizer_scene_gles3.cpp /TP /std:c++14 /nologo /O2 /Z7 /MT /Gd /GR /utf-8 /W3 /wd4267 /wd4244 /wd4305 /wd4018 /wd4800 /EHsc /WX /DDEBUG_ENABLED /DWINDOWS_ENABLED /DOPENGL_ENABLED /DWASAPI_ENABLED /DWINMIDI_ENABLED /DTYPED_METHOD_BIND /DWIN32 /DMSVC /DWINVER=0x0601 /D_WIN32_WINNT=0x0601 /DNOMINMAX /D_WIN64 /DPTRCALL_ENABLED /DTOOLS_ENABLED /DGDSCRIPT_ENABLED /DMINIZIP_ENABLED /DZSTD_STATIC_LINKING_ONLY /DGLAD_ENABLED /DGLES_OVER_GL /DFREETYPE_ENABLED /DSVG_ENABLED /Ithirdparty\nanosvg /Ithirdparty\freetype\include /Ithirdparty\libpng /Ithirdparty\glad /Ithirdparty\zstd /Ithirdparty\zlib /Iplatform\windows /I.
rasterizer_scene_gles3.cpp
D:\a\godot\godot\drivers\gles3\rasterizer_scene_gles3.cpp(2222) : error C2220: the following warning is treated as an error
D:\a\godot\godot\drivers\gles3\rasterizer_scene_gles3.cpp(2222) : warning C4750: 'void __cdecl RasterizerSceneGLES3::_setup_light(struct RasterizerSceneGLES3::RenderList::Element * __ptr64,class Transform const & __ptr64) __ptr64': function with _alloca() inlined into a loop
```

File: drivers/gles3/rasterizer_scene_gles3.cpp [hunk 6]
```diff
@@ -1834,15 +1834,16 @@ void RasterizerSceneGLES3::_render_geometry(RenderList::Element *e) {
 
 void RasterizerSceneGLES3::_setup_light(RenderList::Element *e, const Transform &p_view_transform) {
 
-	int omni_indices[16];
+	int maxobj = state.max_forward_lights_per_object;
+
```
Comment (akien-mga) [inline drivers/gles3/rasterizer_scene_gles3.cpp:1838]: Trailing whitespace.

File: drivers/gles3/rasterizer_scene_gles3.cpp [hunk 7]
```diff
@@ -1834,15 +1834,16 @@ void RasterizerSceneGLES3::_render_geometry(RenderList::Element *e) {
 
 void RasterizerSceneGLES3::_setup_light(RenderList::Element *e, const Transform &p_view_transform) {
 
-	int omni_indices[16];
+	int maxobj = state.max_forward_lights_per_object;
+	
+	int *omni_indices = (int *)alloca(maxobj * sizeof(int));
 	int omni_count = 0;
-	int spot_indices[16];
+
```
Comment (akien-mga) [inline drivers/gles3/rasterizer_scene_gles3.cpp:1841]: Trailing whitespace.

File: drivers/gles3/rasterizer_scene_gles3.cpp [hunk 8]
```diff
@@ -1883,6 +1882,7 @@ void RasterizerSceneGLES3::_setup_light(RenderList::Element *e, const Transform
 		glUniform1iv(state.scene_shader.get_uniform(SceneShaderGLES3::SPOT_LIGHT_INDICES), spot_count, spot_indices);
 	}
 
+
```
Comment (akien-mga) [inline drivers/gles3/rasterizer_scene_gles3.cpp:1885]: Style issue, there shouldn't be two consecutive empty lines.

File: drivers/gles3/rasterizer_scene_gles3.cpp [hunk 9]
```diff
@@ -1834,15 +1834,14 @@ void RasterizerSceneGLES3::_render_geometry(RenderList::Element *e) {
 
 void RasterizerSceneGLES3::_setup_light(RenderList::Element *e, const Transform &p_view_transform) {
 
-	int omni_indices[16];
+	int maxobj = state.max_forward_lights_per_object;
+	int *omni_indices = (int *)alloca(maxobj * sizeof(int));
```
Comment (akien-mga) [inline drivers/gles3/rasterizer_scene_gles3.cpp:1838]: MSVC doesn't seem to like `alloca` here:
```
D:\a\godot\godot\drivers\gles3\rasterizer_scene_gles3.cpp(2219) : error C2220: the following warning is treated as an error
D:\a\godot\godot\drivers\gles3\rasterizer_scene_gles3.cpp(2219) : warning C4750: 'void __cdecl RasterizerSceneGLES3::_setup_light(struct RasterizerSceneGLES3::RenderList::Element * __ptr64,class Transform const & __ptr64) __ptr64': function with _alloca() inlined into a loop
```

Seems like you might have to remove `_FORCE_INLINE_` for it to compile.

BTW, would it make sense to use something smaller than `int` to allocate less memory (e.g. `uint8_t` or `uint16_t`)? Or it doesn't matter?

**Comment by EzraT** [comment]
Fantastic! Thank you!

File: drivers/gles3/rasterizer_scene_gles3.cpp (hunk 9 above)
Comment (NHodgesVFX) [reply drivers/gles3/rasterizer_scene_gles3.cpp:1838]: it would use less memory but even with 1000 lights it only uses around 0.005mb, so I don't really think it matters either way

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

File: drivers/gles3/rasterizer_scene_gles3.cpp (hunk 9 above)
Comment (lawnjelly) [reply drivers/gles3/rasterizer_scene_gles3.cpp:1838]: `int` is probably fine here as the max size is 1024 * sizeof (int), but in general when using the stack, akien is absolutely right, because the stack can be of very limited size, platform dependent, and how much room you have 'depends'. It absolutely is not the same as general memory.

See e.g. here, suggesting historical figures of 8K being possible on old mobile:
https://stackoverflow.com/questions/1859327/android-stack-size

So 1024 * 8 byte int = 8K, blown stack. Whereas 1024 * 2 byte uint16_t = slightly safer, but still likely to blow such a small stack.

More realistically, modern times I would expect around a 1 meg+, especially on desktop:
https://pspdfkit.com/blog/2019/android-large-memory-requirements/

But this is not just for your function.

In short, alloca is great but make sure you know what you are doing, and beware of anything that might be called recursively. This kind of thing is why the compiler refuses to compile with a FORCE_INLINE.
